### DISCUSSION TITLE
Component for Grasshopper

### DISCUSSION THREAD

**Discussion by ktrn4896** [idea-statica/ideastatica-public#447] [Rhino/Grasshopper] [answered]
Hi! My name is Ekaterina and I am doing my Master's thesis about bridge steel connections. Even though IDEA StatiCa Grasshopper plugin is quite developed, I am lacking some functionality and I am willing to create my own components to complement existing ones. I am using C# and `IdeaRS.OpenModel` NuGet package for this.
But I am facing a type casting problem: 
for example I need to take IOM as an input, add something to it and then - send IOM back to the output (the same applicable to materials, members, etc.). To do so I need to cast some `GH_goo` inheritor class to the `OpenModel` class. Normally something like this:
`OpenModel model = ((GH_Goo<OpenModell>)inputIOM).Value `
should work, but it does not. 
Am I missing something? Any hint how to approach this problem? Any NuGet packages/examples/documentation?

**Comment by nathanlukeIDEA** [comment]
Hi @ktrn4896,

First I would like to ask what you are trying to achieve, what you are describing is something that is likely already achievable or if not I can add functionality quite quickly to make it possible if it is a small request / possible.

For example, you can already add things to an existing open model using the plugin. This might help:
https://developer.ideastatica.com/docs/extensions/grasshopper/grasshopper_iom_model_components.html#model-compilation

The objects passed around in the plugin are already wrapped up as Goo and they are also (mostly) wrapped on a rhino wrapper layer. Implementing the unwrapping in your project may be a little difficult as you would need to reference all the plugin .dlls. The easiest way to do it would be to pass the output XML from one of the plugin's components and have your component take a string input (use Inspect XML component). You can use IdeaRS.OpenModel to deserialise it directly to the OpenModel class and then you can work directly with the open model as you wish. 

To import it back into the plug you would have to save it to disk first and then re-import it. I could add an inline serialiser to the plugin if needed. 

Hope this helps. 

Cheers
Nathan

**Comment by ktrn4896** [comment]
Hi @nathanlukeIDEA, 
First of all, thank you very much for the fast and detailed response! 

In my thesis I am dealing with kind of exotic examples (let's say 100 years old bridge), so, I may need something which is not relevant/valuable for majority of users and so, not your first priority. For example, I may need:

1.  To create a material with modified mechanical properties (e.g., create standard material, but set fy to different value - see [Steel Materials](https://developer.ideastatica.com/docs/iom/iom_materials.html#steel-materials) 
2. To create general shape cross section from GH geometry - see [Creating a General Welded Cross-Section](https://developer.ideastatica.com/docs/iom/iom_cross_sections.html#creating-a-general-welded-cross-section)
3. Modify code settings parameters - see [Code settings](https://github.com/idea-statica/ideastatica-public/discussions/234#discussioncomment-4795038)
4. To get any object type instance from a model by name. As I understand, `CreateReference (IOM)` component should solve this problem, but only a few components can take `Reference Element` instead of class instance as an input. I understand  that to get an instance from the model I need to use the same `OpenModel ` instance for the whole script and it is a bit against your ideology, but still I may need it :) 
5. ... and etc. 

All these features are available via code. But GH is kind of industry standard for now and it has a lot of other features and connections, which I do not want to abandon, going fully to a code editor. In every of these examples I need to cast not only `OpenModel` but other classes from/to `Goo` wrapper, so, while your suggestion to use .xml serialization will work for `OpenModel` object, I suppose, it won't work for materials/cross sections and etc. 

So, an opportunity to have an access to full OpenModel NuGet functionality when writing a custom component for GH, would be a big advantage of your program. I understand that it may require extra work for you for preparing a NuGet, documentation and examples, so not asking for it now :) 

And by the way, if I link `IdeaRS.OpenModel.dll` to c# component - I am getting an error `"Error (CS0012): The type 'System.Object' is defined in an assembly that is not referenced. You must add a reference to assembly 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'."` and I do not understand how to fix it. If I compile my custom component to deal with IOM using Visual Studio Code, I do not have the same error.

**Reply by nathanlukeIDEA** [reply, marked as answer]
Hi @ktrn4896, 

Firstly, this is how you can fix the IdeaRS.OpenModel referencing in the c# component.
https://discourse.mcneel.com/t/running-netstandard-2-0-dll-within-a-c-scripting-component/146522

I have attached an example of creating a material using IOM (inside a c-sharp component) and then passing it to one of the Idea Grasshopper component using a c# component. You can reference all the .dll's from the IDEA StatiCa plugin directory. You will need to copy the IdeaStatiCaGrasshopper.gha somewhere and rename the extension to .dll. This is so you can reference the Idea Goo classes. This should allow you to create custom materials and cross-sections. Make sure that any objects you create this way have a unique ID and Name and just be aware that some things may not work as expected because this is not really how it was intended to be used.
[csharp_material_from_code.zip](https://github.com/idea-statica/ideastatica-public/files/14325396/csharp_material_from_code.zip)

The code setup will not work like this (as it is not implemented on the rhino wrapper side) - you will need to change this through once the connection is created through the Connection API components or within the program itself, unless you do the modification at the end, once the model is fully compiled.  

Hopefully this helps. 

Cheers
Nathan

**Reply by ktrn4896** [reply]
Hi, @nathanlukeIDEA, 

Thank you very much! Yes, this helps a lot! Will try to implement my wishes :) 

Best regards, 
Ekaterina

**Reply by mpospisil** [reply]
Hi Ekaterina,

Your master's thesis looks very interesting.
It could be a nice example of usage our API. If you want, you can contact me directly  by email to discuss how to fullfill your requirements.

martin.pospisil@ideastatica.com

**Reply by ktrn4896** [reply]
Hi, Martin, 

Thank you very much, nice to hear!
I will try to implement my ideas with all what Nathan explained to me (I did couple of tests and it looks promising). 
I will contact you, if any questions/new ideas. 

Best regards,
Ekaterina
